scipy/sparse/_construct.py

```diff
@@ -681,7 +707,60 @@ def bmat(blocks, format=None, dtype=None):
            [0, 0, 7]])
 
     """
+    blocks = np.asarray(blocks, dtype='object')
+    if any(isinstance(b, sparray) for b in blocks.flat):
+        return _block_array(blocks, format, dtype, coo_array)
+    else:
+        return _block_array(blocks, format, dtype, coo_matrix)
 
+
+def block_array(blocks, *, format=None, dtype=None):
```

Comment:
I believe the idea was to rename the construction functions using e.g. `eye_array` when they return arrays. I did think about using `block` to match other libraries. But I decided it was a poor choice of name because the word block doesn't say what we are doing/returning/wanting and it means too many things. In english "block" can be a verb, noun, adjective which can be confusing. It's an adjective instead of a noun or verb in this case. `block_array` at least makes that clear. And it describes what we expect to get from this.

Reply:
Typical famous names for these are `spdiags`, and other `spXXXX` stuff following mathworks. And provides a better separation from the dense ones.  `sp_concat` might be an option but a very loose suggestion, please ignore freely.

Reply:
Given that we already have `scipy.sparse.hstack` and `scipy.sparse.vstack`, I think that `scipy.sparse.block` makes sense here. We do have the `diags`/`diags_array` situation at present, but eventually the plan is to end up renaming `diags_array` back to `diags`, so if we don't need the _array suffix here then I'm happy to avoid it from the start.

I'd like to stay away from Matlab-isms if possible, especially if we can instead align more closely with the numpy-style naming conventions.

Reply:
I changed the name to `block` instead of `block_array`.

Reply:
Thanks for the change!